src/nvim/tui/input.c

```diff
@@ -282,27 +289,56 @@ static bool handle_focus_event(TermInput *input)
 
 static bool handle_bracketed_paste(TermInput *input)
 {
-  if (rbuffer_size(input->read_stream.buffer) > 5 &&
-      (!rbuffer_cmp(input->read_stream.buffer, "\x1b[200~", 6) ||
-       !rbuffer_cmp(input->read_stream.buffer, "\x1b[201~", 6))) {
-    bool enable = *rbuffer_get(input->read_stream.buffer, 4) == '0';
-    // Advance past the sequence
-    rbuffer_consumed(input->read_stream.buffer, 6);
-    if (input->paste_enabled == enable) {
-      return true;
+  RBuffer *rbuf = input->read_stream.buffer;
+  if (rbuffer_size(rbuf) > 5
+      && (!rbuffer_cmp(rbuf, "\x1b[200~", 6)
+          || !rbuffer_cmp(rbuf, "\x1b[201~", 6))) {
+    bool enable = *rbuffer_get(rbuf, 4) == '0';
+
+    if (enable && input->paste_started) {
+      // Received a bogus "paste start" after paste was already started.
+      return false;
+    }
+
+    rbuffer_consumed(rbuf, 6);  // Advance past the sequence
+    input->paste_started = enable;
+    if (enable) {
+      loop_schedule(&loop, event_create(1, apply_pastepre, 0));
+    } else {
+      enqueue_input(input, PASTEPOST_KEY, sizeof(PASTEPOST_KEY) - 1);
```

Comment:
@tarruda / @oni-link, Wondering if you can help my understanding. `loop_schedule()` on the "start paste" case (line 306) works well, because it puts the event on the main queue before any input gets processed. But for the "stop paste" case, I had to enqueue this `PASTEPOST_KEY` in order to guarantee that all paste data was processed before queuing the `PastePost` event. 

 Do we currently have any better mechanism for placing an event at a precise location in the input stream (I'm guessing the answer is "no", hence why we still have these pseudokeys)? I also tried using a mutex, but that had problems and isn't really desirable anyways.

I am wondering if it would be a good idea to provide some internal generic mechanism for placing `K_EVENT` in the input stream and associating it with a callback.